layer/unittest/trace_event_log_tests.cc

```diff
@@ -27,8 +31,167 @@ using ::testing::IsEmpty;
+
+std::string ValueToJsonString(int64_t value) { return std::to_string(value); }
+
+std::string ValueToJsonString(const std::vector<int64_t> &values) {
+  std::stringstream json_str;
+  json_str << "[";
+  size_t e = values.size();
+  for (size_t i = 0; i != e; ++i) {
+    const char *delimiter = i < e - 1 ? "," : "";
+    json_str << "\"0x" << std::hex << values[i] << "\"" << delimiter;
+  }
+  json_str << "]";
+  return json_str.str();
+}
+
+// Converts the duration to milliseconds, the defualt time unit in the `Trace
+// Event` format.
+std::string ValueToJsonString(Duration value) {
+  return std::to_string(value.ToMilliseconds());
+}
+
+// Converts the timestamp to milliseconds, the defualt time unit in the `Trace
+// Event` format.
+std::string ValueToJsonString(TimestampClock::time_point value) {
+  return std::to_string(ToUnixMillis(value));
+}
+
+// Appends all given attributes, including the type-specific attribute, to the
+// stream.
+void TraceArgsToJsonString(const std::vector<Attribute *> &args,
+                           std::ostringstream &json_str) {
+  json_str << ", " << std::quoted("args") << " : { ";
+  for (size_t i = 0, e = args.size(); i < e; ++i) {
+    json_str << std::quoted(args[i]->GetName()) << " : ";
+    switch (args[i]->GetValueType()) {
+      case kHashAttribute: {
+        json_str << "\"0x" << std::hex << args[i]->cast<HashAttr>()->GetValue()
+                 << "\"";
+        break;
+      }
+      case ValueType::kTimestamp: {
+        json_str << ValueToJsonString(
+            args[i]->cast<TimestampAttr>()->GetValue());
+        break;
+      }
+      case ValueType::kDuration: {
+        json_str << ValueToJsonString(
+            args[i]->cast<DurationAttr>()->GetValue());
+        break;
+      }
+      case ValueType::kBool: {
+        json_str << ValueToJsonString(args[i]->cast<BoolAttr>()->GetValue());
+        break;
+      }
+      case ValueType::kInt64: {
+        json_str << ValueToJsonString(args[i]->cast<Int64Attr>()->GetValue());
+        break;
+      }
+      case ValueType::kString: {
+        json_str << std::quoted(args[i]->cast<StringAttr>()->GetValue());
+        break;
+      }
+      case ValueType::kVectorInt64: {
+        json_str << ValueToJsonString(
+            args[i]->cast<VectorInt64Attr>()->GetValue());
+        break;
+      }
+      case ValueType::kTraceEvent:
+        break;
+    }
+  }
+  json_str << " }";
+}
+
+// Appends the start timestamp and the duration of a `TraceEventAttr` to the
+// given stream.
+void CompleteEventToString(TimestampAttr timestamp,
+                           const TraceEventAttr *trace_event,
+                           std::ostringstream &json_stream) {
+  const std::vector<Attribute *> &args = trace_event->GetArgs();
+  assert(args.size() && "CompleteEvent's args must not be empty.");
+
+  auto it = std::find_if(args.begin(), args.end(), [](Attribute *attr) {
+    return attr->isa<DurationAttr>();
+  });
```

Comment:
Idea: what do you think about adding a helper function to `TraceEventAttr` to retrieve attributes based on types? Something like:
```c++
auto *duration = trace_event->GetArg<DurationAttr>();
assert(duration && "Duration not found");
```

We could also add a similar helper to `Event`.

Reply:
That's a good idea. Maybe I can overload it for a `string`  input as well to scan based on the name. This would be useful for `scope`.

Reply:
👍